### DISCUSSION TITLE
Optional uploadfile official sample case didn't run

### DISCUSSION THREAD

**Discussion by ninalty** [fastapi/fastapi#13070] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from fastapi import FastAPI, File, UploadFile

app = FastAPI()


# @app.post("/files/")
# async def create_file(file: Annotated[bytes | None, File()] = None):
#     if not file:
#         return {"message": "No file sent"}
#     else:
#         return {"file_size": len(file)}


@app.post("/uploadfile/")
async def create_upload_file(file: UploadFile | None = None):
    if not file:
        return {"message": "No upload file sent"}
    else:
        return {"filename": file.filename}
```

### Description

This is the sample code from official website.https://fastapi.tiangolo.com/tutorial/request-files/#optional-file-upload

The first case works (using the bytes). The 2nd case with UploadFile doesn't work.

### Operating System

macOS

### FastAPI Version

0.115.6

### Pydantic Version

2.9.2

### Python Version

3.11.4

**Comment by Knighthawk-Leo** [comment]
@ninalty 
I think it's the issue from the swagger. It is working fine with postman your Code snippet is correct 

![Screenshot from 2024-12-14 22-47-24](https://github.com/user-attachments/assets/ecbb3ed3-e9b2-456c-88d9-8e70a1f711d3)

**Reply by YuriiMotov** [reply]
Not sure this issue is from Swagger.

Openapi schemas for both cases are identical and Swagger sends identical requests.

Moreover, problem is reproducible with curl (only URL has been changed in the requests):
```
curl -X 'POST' 'http://127.0.0.1:8000/files/' -H 'accept: application/json' -H 'Content-Type: multipart/form-data' -F 'file='
# {"message":"No file sent"}

curl -X 'POST' 'http://127.0.0.1:8000/uploadfile/' -H 'accept: application/json' -H 'Content-Type: multipart/form-data' -F 'file='
# {"detail":[{"type":"value_error","loc":["body","file"],"msg":"Value error, Expected UploadFile, received: <class 'str'>","input":"","ctx":{"error":{}}}]}
```

**Comment by yvallois** [comment]
If you replace your first endpoint with 
```python
@app.post("/files/")
async def create_file(file: Annotated[bytes | None, File()] = None):
    if file is None:
        return {"message": "No file sent"}
    else:
        return {"file_size": len(file)}
```
you'll see that file is an empty string if you send it via documentation or curl.
@Knighthawk-Leo  in your example postman won't send a file or even Content-Type: multipart/form-data

If we send `curl -X 'POST' 'http://127.0.0.1:8000/files/' -H 'accept: application/json' -H 'Content-Type: multipart/form-data' -F 'key=value'` to the 2 endpoints, we get the expected behavior.
on the other hand, `curl -X 'POST' 'http://127.0.0.1:8000/files/' -H 'accept: application/json' -H 'Content-Type: multipart/form-data' -F 'file=' file` contains an empty string. This doesn't cause a problem for the first endpoint (it accepts a text via postman) but with UploadFile the validator used is `TypeAdapter(Annotated[Union[UploadFile, NoneType], File(None)])` which doesn't accept an empty string.

You can correct this behavior by changing the signature with file: UploadFile | str | None = None

**Reply by YuriiMotov** [reply]
> You can correct this behavior by changing the signature with file: UploadFile | str | None = None

IMHO, not a good idea since it will brake the schema (allow passing string)

**Reply by yvallois** [reply]
You're right, in any case I don't think it's bad behaviour to fastapi here.

**Comment by YuriiMotov** [comment, marked as answer]
This was fixed by https://github.com/fastapi/fastapi/pull/13537 and available in FastAPI [0.123.2](https://github.com/fastapi/fastapi/releases/tag/0.123.2)
